### PR TITLE
[12.0][ADD] - Improve multi-company management for brand model

### PR DESCRIPTION
**Pull request by sbejaoui**
In a multi-company management system, there is a strong link between the company
and its brands. A company can't make sells under a brand that it don't own.
For this reason we add a mixin to be inherited from branded objects that assure the coherence of this link.

We also add an multi-company access rule for brand model.

A new field added to the company to define if the brand field is required or optional by company.

### CODE REVIEW

**Comment by sbejaoui** [comment]
In this PR, I made a refactoring to introduce the abstract model `res.brand.mixin`.

It made more sense for me to have base module `brand` that define the brand model and its views, security and menu. But this left the module `partner_brand` only with the improvement of the report layout.

I propose to rename this  module to `brand_report_layout`. `partner_brand` don't make sense any more after we changed the design of brand model.

I propose to rename the other modules too:
* sale_brand --> brand_sale
* account_brand --> brand_account

@pedrobaeza proposed to use `company_brand` instead of `brand` to clarify the difference with `product_brand`.

It ok to me too.

@max3903 , @osi-scampbell , @pedrobaeza , @sbidoul

**Comment by max3903** [comment]
@sbejaoui Can you split the pull request to only have `brand` in this one?

**Comment by sbejaoui** [comment]
> @sbejaoui Can you split the pull request to only have `brand` in this one?

I will split my work in different commits (one commit by module), but in the PR we need the changes on all the modules as they are depending on the mixin.

**Comment by max3903** [comment]
@sbejaoui What about:

* 1 pull request for brand
* 1 pull request with separate commit for brand_*

**Comment by sbejaoui** [comment]
> @sbejaoui What about:
> 
> * 1 pull request for brand
> * 1 pull request with separate commit for brand_*

We will not be able to test them in runbot as there is a strong dependency bewteen modules

File: brand/readme/DESCRIPTION.rst [hunk 1]
```diff
@@ -0,0 +1 @@
+This is a base addon for brand modules. It adds the brand object and its menu and define an abstract model to be inherited from branded objects
```
Comment (sbidoul) [inline brand/readme/DESCRIPTION.rst:1]: Perhaps add a `USAGE.rst` pointing to the simplest module that uses the mixing as an example? Also say a word about it's configuration, which you can copy in the PR description to help reviewers.

File: brand/views/res_config_settings.xml [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+                    <div class="col-12 col-lg-6 o_setting_box">
+                        <div class="o_setting_right_pane">
+                            <span class="o_form_label">Brand</span>
+                            <span class="fa fa-lg fa-building-o"
+                                  title="Values set here are company-specific."
+                                  aria-label="Values set here are company-specific."
+                                  groups="base.group_multi_company"
+                                  role="img"/>
+                            <div class="text-muted">
+                                Set the brand use level for the branded objects.
```
Comment (sbidoul) [inline brand/views/res_config_settings.xml]:
```suggestion
                                Configure branding of customer facing business documents (SO, invoices, etc).
```

File: brand/models/res_company.py [hunk 3]
```diff
@@ -0,0 +1,23 @@
+# Copyright 2019 ACSONE SA/NV
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+
+from odoo import fields, models
+
+BRAND_USE_LEVEL_SELECTION = [
+    ('no', 'No Use'),
```
Comment (sbidoul) [inline brand/models/res_company.py]:
```suggestion
    ('no', 'Do not use brands on business document'),
```

File: brand/models/res_company.py [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+    ('optional', 'Optional'),
+    ('required', 'Required'),
```
Comment (sbidoul) [inline brand/models/res_company.py]:
```suggestion
    (BRAND_USE_LEVEL_REQUIRED_LEVEL, 'Required'),
```

File: brand/models/res_brand_mixin.py [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+    def _check_brand_company_id(self):
+        for rec in self:
+            if (
+                rec.brand_id.company_id
+                and rec.brand_id.company_id != rec.company_id
+            ):
+                raise ValidationError(
+                    _(
+                        "Brand company is different then "
+                        "the selected company"
```
Comment (sbidoul) [inline brand/models/res_brand_mixin.py]: Brand must match document company for {rec.display_name}

File: brand/models/res_brand_mixin.py [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+                    )
+                )
+
+    @api.onchange('brand_id')
+    def _onchange_brand_id(self):
+        if 'company_id' in self._fields:
```
Comment (sbidoul) [inline brand/models/res_brand_mixin.py]: is this test still necessary?

File: brand/models/res_brand_mixin.py [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+                )
+                orm.setup_modifiers(
+                    elem,
+                    self.fields_get(['brand_use_level'])['brand_use_level'],
+                )
+                node.addprevious(elem)
+                node.set(
+                    'attrs',
+                    '{"invisible": '
+                    '[("brand_use_level", "=", "no")], '
```
Comment (sbidoul) [inline brand/models/res_brand_mixin.py]: use constant for "no" too

**Review by sbidoul — COMMENTED** [review]
A few minor comments

File: brand/models/res_brand_mixin.py [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+
+    @api.constrains('brand_id', 'company_id')
+    def _check_brand_company_id(self):
+        for rec in self:
+            if (
+                rec.brand_id.company_id
+                and rec.brand_id.company_id != rec.company_id
+            ):
+                raise ValidationError(
+                    _("Brand must match document company for %s")
```
Comment (sbidoul) [inline brand/models/res_brand_mixin.py]:
```suggestion
                    _("Brand company must match document company for %s")
```

**Review by sbidoul — APPROVED** [review]
Looks good, great job. This adds flexibility while preserving compatibility for the `partner_brand` module..

**Review by osi-scampbell — APPROVED** [review]

**Pull request merged**